src/python/pants/engine/README.md

```diff
@@ -17,63 +17,74 @@ Once the engine is instantiated with a valid set of `@rule`s, a caller can synch
+would thus need that `Param` in order to run. To see which `Params` the engine needs to run each
+`@rule`, refer to the `Visualization` section below.
 
-The return value of an `@rule` for a particular `Subject` is known as a `Product`. At some level,
-you can think of (`subject_value`, `product_type`) as a "key" that uniquely identifies a particular
-Product value and `@rule` execution.
+Any hashable type with useful equality may be used as a `Param`, and additional `Params` can be
+provided to an `@rule`'s dependencies via `Get` requests (see below). Each `Param` value in a set
+of `Params` is unique by type, so if `@rules` recursively introduce a particular `Param` type,
+there will still only be one value for that type in each `@rule`, but it will change as you move
+deeper into the dependency graph.
+
+The return value of an `@rule` is known as a `Product`. At some level, you can think
+of `(product_type, params_set)` as a "key" that uniquely identifies a particular `Product` value
+and `@rule` execution. If an `@rule` is able to produce a `Product` without consuming any `Params`,
+then the `@rule` will run exactly once, and the value that it produces will be a singleton.
 
 #### Example
 
 As a very simple example, you might register the following `@rule` that can compute a `String`
-Product given a single `Int` input.
+Product given a single `Int` argument.
 
 ```python
-@rule(StringType, [IntType])
+@rule(str, [int])
 def int_to_str(an_int):
-  return '{}'.format(an_int)
+  return str(an_int)
 ```
 
-The first argument to the `@rule` decorator is the Product (ie, return) type for the `@rule`. The
-second argument is a list of parameter selectors that declare the types of the input parameters for
-the `@rule`. In this case, because the Product type is `StringType` and there is one parameter
-selector (for `IntType`), this `@rule` represents a conversion from `IntType` to `StrType`, with no
-other inputs.
+The first argument to the `@rule` decorator is the `Product` (ie, return) type for the `@rule`. The
+second argument is a list of "parameter selectors" that declare the types of the input parameters for
+the `@rule`. In this case, because the `Product` type is `str` and there is one parameter
+selector (for `int`), this `@rule` represents a conversion from `int` to `str`, with no other inputs.
+
+When the engine encounters this `@rule` while compiling the rule graph for `str`-producing-`@rules`,
+it will next go hunting for the dependency `@rule` that can produce an `int` using the fewest number
+of `Params`. For example, if there was an `@rule` that could produce an `int` without consuming any
+`Params` at all (ie, a singleton), then that `@rule` would always be chosen first. If all `@rules` to
+produce `int`s required at least one `Param`, then the engine would next see whether the input `Params`
+contained an `int`, or whether there were any `@rules` that required only one `Param`, then two
```

Comment:
This sentence buries the lede a bit: you start by talking about needing to produce an int from Params, but here you imply that the int itself can be a Param. 

Does the output of some other rule count as a Param, or are Params just the things that are injected into the boundary of the graph "from outside"? 

In other words, if I have `rule1` that takes a Foo and returns a Bar, and then `rule2` that takes a Bar and returns a Baz, is it correct to refer to the Foo as a Param? What about the Bar? 

Sorry to keep banging on this, but this README is going to be extremely useful and important, so it's best to make sure it's crystal clear.